src/query/ee/src/fail_safe/handler.rs

```diff
@@ -0,0 +1,385 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+use std::time::Duration;
+
+use aws_config::meta::region::RegionProviderChain;
+use aws_config::retry::RetryConfig;
+use aws_config::timeout::TimeoutConfig;
+use aws_sdk_s3::config::Credentials;
+use aws_sdk_s3::config::Region;
+use aws_sdk_s3::config::SharedCredentialsProvider;
+use aws_sdk_s3::Client;
+use databend_common_base::base::tokio::io::AsyncReadExt;
+use databend_common_base::base::GlobalInstance;
+use databend_common_catalog::table::Table;
+use databend_common_config::GlobalConfig;
+use databend_common_exception::ErrorCode;
+use databend_common_exception::Result;
+use databend_common_expression::TableSchema;
+use databend_common_meta_app::schema::TableInfo;
+use databend_common_meta_app::storage::StorageParams;
+use databend_common_storages_fuse::io::MetaReaders;
+use databend_common_storages_fuse::FuseTable;
+use databend_enterprise_fail_safe::FailSafeHandler;
+use databend_enterprise_fail_safe::FailSafeHandlerWrapper;
+use databend_storages_common_cache::LoadParams;
+use databend_storages_common_table_meta::meta::CompactSegmentInfo;
+use databend_storages_common_table_meta::meta::Location;
+use log::info;
+use log::warn;
+use opendal::ErrorKind;
+use opendal::Operator;
+
+pub struct RealFailSafeHandler {}
+
+impl RealFailSafeHandler {}
+
+#[async_trait::async_trait]
+impl FailSafeHandler for RealFailSafeHandler {
+    async fn recover_table_data(&self, table_info: TableInfo) -> Result<()> {
+        let storage_params = match &table_info.meta.storage_params {
+            // External or attached table.
+            Some(sp) => sp.clone(),
+            // Normal table.
+            None => {
+                let config = GlobalConfig::instance();
+                config.storage.params.clone()
+            }
+        };
+
+        let fuse_table = FuseTable::do_create(table_info)?;
+
+        let amender = Amender::try_new(storage_params).await?;
+
+        amender.recover_snapshot(fuse_table).await?;
+
+        Ok(())
+    }
+}
+
+impl RealFailSafeHandler {
+    pub fn init() -> Result<()> {
+        let rm = RealFailSafeHandler {};
+        let wrapper = FailSafeHandlerWrapper::new(Box::new(rm));
+        GlobalInstance::set(Arc::new(wrapper));
+        Ok(())
+    }
+}
+
+struct Amender {
+    client: Client,
+    bucket: String,
+    root: String,
+}
+
+impl Amender {
+    async fn try_new(storage_param: StorageParams) -> Result<Self> {
+        // TODO
+        // - replace client with opendal operator
+        // - supports other storage types
+        // when `list_object_versions` or higher level api is supported by opendal
+        if let StorageParams::S3(s3_config) = storage_param {
+            if s3_config.access_key_id.is_empty() {
+                panic!()
```

Comment:
It would better to add detailed error messages

Reply:
:+1:  thanks